gcp-client/src/main/java/com/google/graphite/platforms/plugin/client/ComputeClient.java

```diff
@@ -0,0 +1,541 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.graphite.platforms.plugin.client;
+
+import com.google.api.services.compute.Compute;
+import com.google.api.services.compute.model.AcceleratorType;
+import com.google.api.services.compute.model.AttachedDisk;
+import com.google.api.services.compute.model.DiskType;
+import com.google.api.services.compute.model.Image;
+import com.google.api.services.compute.model.Instance;
+import com.google.api.services.compute.model.InstanceTemplate;
+import com.google.api.services.compute.model.InstancesScopedList;
+import com.google.api.services.compute.model.MachineType;
+import com.google.api.services.compute.model.Metadata;
+import com.google.api.services.compute.model.Network;
+import com.google.api.services.compute.model.Operation;
+import com.google.api.services.compute.model.Region;
+import com.google.api.services.compute.model.Snapshot;
+import com.google.api.services.compute.model.SnapshotList;
+import com.google.api.services.compute.model.Subnetwork;
+import com.google.api.services.compute.model.Zone;
+import com.google.common.base.Strings;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Client for communicating with the Google Compute API
+ *
+ * @see <a href="https://cloud.google.com/compute/">Cloud Engine</a>
+ */
+public class ComputeClient {
+  private Compute compute;
+
+  private static final Logger LOGGER = Logger.getLogger(ComputeClient.class.getName());
+  private static final long SNAPSHOT_TIMEOUT_MILLISECONDS = TimeUnit.MINUTES.toMillis(10);
+
+  public ComputeClient(Compute compute) {
+    setCompute(compute);
+  }
+
+  public static String nameFromSelfLink(String selfLink) {
+    return selfLink.substring(selfLink.lastIndexOf("/") + 1);
+  }
+
+  public static String buildLabelsFilterString(Map<String, String> labels) {
+    StringBuilder sb = new StringBuilder();
+    for (Map.Entry<String, String> l : labels.entrySet()) {
+      sb.append("(labels." + l.getKey() + " eq " + l.getValue() + ") ");
+    }
+    return sb.toString().trim();
+  }
+
+  public static List<Metadata.Items> mergeMetadataItems(
+      List<Metadata.Items> winner, List<Metadata.Items> loser) {
+    if (loser == null) {
+      loser = new ArrayList<>();
+    }
+
+    // Remove any existing metadata that has the same key(s) as what we're trying to update/append
+    for (Metadata.Items existing : loser) {
+      boolean duplicate = false;
+      for (Metadata.Items newItem : winner) { // Items to append
+        if (existing.getKey().equals(newItem.getKey())) {
+          duplicate = true;
+        }
+      }
+      if (!duplicate) {
+        winner.add(existing);
```

Comment:
Would rather not modify the parameter list, as it could have unintended side-effects. Suggest creating a local copy and modifying that instead.

Reply:
This was copied and pasted from the GCE plugin except for the package, so I'm glad we can look at this in a fresh light with this PR.

Reply:
Done